File: sandbox/numerical/src/mesh.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from math import hypot
from typing import Iterable


class CellType(str, Enum):
    Triangle = "triangle"
    Quadrilateral = "quadrilateral"


@dataclass(frozen=True)
class Node:
    id: str
    x: float
    y: float


@dataclass(frozen=True)
class Cell:
    id: str
    cell_type: CellType
    node_ids: tuple[str, ...]

    @property
    def node_count(self) -> int:
        return len(self.node_ids)

    @property
    def edge_count(self) -> int:
        return len(self.node_ids)


@dataclass(frozen=True)
class Edge:
    id: str
    node_ids: tuple[str, str]
    left_cell: str | None
    right_cell: str | None
    normal: tuple[float, float]
    length: float
    midpoint: tuple[float, float]


@dataclass(frozen=True)
class EdgeSpec:
    id: str
    node_ids: tuple[str, str]
    left_cell: str | None
    right_cell: str | None


@dataclass(frozen=True)
class Mesh:
    nodes: tuple[Node, ...]
    cells: tuple[Cell, ...]
    edges: tuple[Edge, ...]


def build_mixed_minimal_mesh() -> Mesh:
    nodes = (
        Node("N0", 0.0, 0.0),
        Node("N1", 1.0, 0.0),
        Node("N2", 2.0, 0.0),
        Node("N3", 0.0, 1.0),
        Node("N4", 1.0, 1.0),
        Node("N5", 2.0, 1.0),
    )
    cells = (
        Cell("C0", CellType.Quadrilateral, ("N0", "N1", "N4", "N3")),
        Cell("C1", CellType.Triangle, ("N1", "N2", "N4")),
        Cell("C2", CellType.Triangle, ("N2", "N5", "N4")),
    )
    edge_specs = (
        EdgeSpec("E0", ("N0", "N1"), None, "C0"),
        EdgeSpec("E1", ("N1", "N4"), "C0", "C1"),
        EdgeSpec("E2", ("N1", "N2"), None, "C1"),
        EdgeSpec("E3", ("N2", "N4"), "C1", "C2"),
        EdgeSpec("E4", ("N4", "N3"), "C0", None),
        EdgeSpec("E5", ("N0", "N3"), None, "C0"),
        EdgeSpec("E6", ("N2", "N5"), None, "C2"),
        EdgeSpec("E7", ("N5", "N4"), "C2", None),
    )
    return build_mesh(nodes, cells, edge_specs)


def build_quad_only_control_mesh() -> Mesh:
    nodes = _control_nodes()
    cells = (
        Cell("C0", CellType.Quadrilateral, ("N0", "N1", "N4", "N3")),
        Cell("C1", CellType.Quadrilateral, ("N1", "N2", "N5", "N4")),
    )
    return build_mesh(nodes, cells)


def build_tri_only_control_mesh() -> Mesh:
    nodes = _control_nodes()
    cells = (
        Cell("C0", CellType.Triangle, ("N0", "N1", "N3")),
        Cell("C1", CellType.Triangle, ("N1", "N4", "N3")),
        Cell("C2", CellType.Triangle, ("N1", "N2", "N4")),
        Cell("C3", CellType.Triangle, ("N2", "N5", "N4")),
    )
    return build_mesh(nodes, cells)


def build_mesh(
    nodes: Iterable[Node],
    cells: Iterable[Cell],
    edge_specs: Iterable[EdgeSpec] | None = None,
) -> Mesh:
    node_tuple = tuple(nodes)
    cell_tuple = tuple(cells)
    node_lookup = {node.id: node for node in node_tuple}
    _validate_cells(cell_tuple, node_lookup)
    edges = _build_edges(cell_tuple, node_lookup, tuple(edge_specs) if edge_specs else None)
    return Mesh(node_tuple, cell_tuple, edges)


def cell_area(cell: Cell, nodes: dict[str, Node]) -> float:
    polygon = [nodes[node_id] for node_id in cell.node_ids]
    twice_area = 0.0
    for index in range(cell.node_count):
        current = polygon[index]
        following = polygon[(index + 1) % cell.node_count]
        twice_area += current.x * following.y - following.x * current.y
    return abs(twice_area) * 0.5


def cell_center(cell: Cell, nodes: dict[str, Node]) -> tuple[float, float]:
    polygon = [nodes[node_id] for node_id in cell.node_ids]
    signed_area_twice = 0.0
    centroid_x = 0.0
    centroid_y = 0.0
    for index in range(cell.node_count):
        current = polygon[index]
        following = polygon[(index + 1) % cell.node_count]
        cross = current.x * following.y - following.x * current.y
        signed_area_twice += cross
        centroid_x += (current.x + following.x) * cross
        centroid_y += (current.y + following.y) * cross
    if signed_area_twice == 0.0:
        raise ValueError(f"cell {cell.id} has zero area")
    return (centroid_x / (3.0 * signed_area_twice), centroid_y / (3.0 * signed_area_twice))


def to_reference_dict(mesh: Mesh) -> dict[str, object]:
    node_lookup = {node.id: node for node in mesh.nodes}
    return {
        "nodes": [node.__dict__ for node in mesh.nodes],
        "cells": [
            {
                "id": cell.id,
                "cell_type": cell.cell_type.value,
                "node_ids": list(cell.node_ids),
                "node_count": cell.node_count,
                "edge_count": cell.edge_count,
                "area": cell_area(cell, node_lookup),
                "center": list(cell_center(cell, node_lookup)),
            }
            for cell in mesh.cells
        ],
        "edges": [
            {
                "id": edge.id,
                "node_ids": list(edge.node_ids),
                "left_cell": edge.left_cell,
                "right_cell": edge.right_cell,
                "normal": list(edge.normal),
                "length": edge.length,
                "midpoint": list(edge.midpoint),
            }
            for edge in mesh.edges
        ],
    }


def _control_nodes() -> tuple[Node, ...]:
    return (
        Node("N0", 0.0, 0.0),
        Node("N1", 1.0, 0.0),
        Node("N2", 2.0, 0.0),
        Node("N3", 0.0, 1.0),
        Node("N4", 1.0, 1.0),
        Node("N5", 2.0, 1.0),
    )


def _validate_cells(cells: tuple[Cell, ...], nodes: dict[str, Node]) -> None:
    for cell in cells:
        if cell.cell_type is CellType.Triangle and cell.node_count != 3:
            raise ValueError(f"triangle {cell.id} must have 3 nodes")
        if cell.cell_type is CellType.Quadrilateral and cell.node_count != 4:
            raise ValueError(f"quadrilateral {cell.id} must have 4 nodes")
        if len(set(cell.node_ids)) != cell.node_count:
            raise ValueError(f"cell {cell.id} has duplicate nodes")
        for node_id in cell.node_ids:
            if node_id not in nodes:
                raise ValueError(f"cell {cell.id} references unknown node {node_id}")
        if cell_area(cell, nodes) == 0.0:
            raise ValueError(f"cell {cell.id} has zero area")


def _build_edges(
    cells: tuple[Cell, ...],
    nodes: dict[str, Node],
    edge_specs: tuple[EdgeSpec, ...] | None,
) -> tuple[Edge, ...]:
    edge_cells: dict[frozenset[str], list[tuple[str, tuple[str, str]]]] = {}
    cell_lookup = {cell.id: cell for cell in cells}
    for cell in cells:
        for index in range(cell.edge_count):
            start = cell.node_ids[index]
            end = cell.node_ids[(index + 1) % cell.edge_count]
            key = frozenset((start, end))
            edge_cells.setdefault(key, []).append((cell.id, (start, end)))

    specs = edge_specs if edge_specs is not None else _default_edge_specs(edge_cells)
    _validate_edge_specs(specs, edge_cells)
    edges: list[Edge] = []
    for spec in specs:
        start = nodes[spec.node_ids[0]]
        end = nodes[spec.node_ids[1]]
        dx = end.x - start.x
        dy = end.y - start.y
        length = hypot(dx, dy)
        if length == 0.0:
            raise ValueError(f"edge {spec.id} has zero length")
        midpoint = ((start.x + end.x) * 0.5, (start.y + end.y) * 0.5)
        normal = _oriented_normal(spec, cell_lookup, nodes, length, dx, dy, midpoint)
        edges.append(
            Edge(
                spec.id,
                spec.node_ids,
                spec.left_cell,
                spec.right_cell,
                normal,
                length,
                midpoint,
            )
        )
    return tuple(edges)


def _oriented_normal(
    spec: EdgeSpec,
    cells: dict[str, Cell],
    nodes: dict[str, Node],
    length: float,
    dx: float,
    dy: float,
    midpoint: tuple[float, float],
) -> tuple[float, float]:
    normal = (dy / length, -dx / length)
    if spec.left_cell is None:
        return _normal_toward_cell(spec.right_cell, cells, nodes, midpoint, normal)
    if spec.right_cell is None:
        return _normal_away_from_cell(spec.left_cell, cells, nodes, midpoint, normal)
    return _normal_from_cell_to_cell(spec.left_cell, spec.right_cell, cells, nodes, midpoint, normal)


def _normal_toward_cell(
    cell_id: str | None,
    cells: dict[str, Cell],
    nodes: dict[str, Node],
    midpoint: tuple[float, float],
    normal: tuple[float, float],
) -> tuple[float, float]:
    if cell_id is None:
        raise ValueError("boundary edge must reference one cell")
    if _cell_side(cell_id, cells, nodes, midpoint, normal) < 0.0:
        return (-normal[0], -normal[1])
    return normal


def _normal_away_from_cell(
    cell_id: str,
    cells: dict[str, Cell],
    nodes: dict[str, Node],
    midpoint: tuple[float, float],
    normal: tuple[float, float],
) -> tuple[float, float]:
    if _cell_side(cell_id, cells, nodes, midpoint, normal) > 0.0:
        return (-normal[0], -normal[1])
    return normal


def _normal_from_cell_to_cell(
    left_cell: str,
    right_cell: str,
    cells: dict[str, Cell],
    nodes: dict[str, Node],
    midpoint: tuple[float, float],
    normal: tuple[float, float],
) -> tuple[float, float]:
    if _cell_side(right_cell, cells, nodes, midpoint, normal) < 0.0:
        normal = (-normal[0], -normal[1])
    if _cell_side(left_cell, cells, nodes, midpoint, normal) > 0.0:
        raise ValueError("edge normal does not separate left and right cells")
    return normal


def _cell_side(
    cell_id: str,
    cells: dict[str, Cell],
    nodes: dict[str, Node],
    midpoint: tuple[float, float],
    normal: tuple[float, float],
) -> float:
    center = cell_center(cells[cell_id], nodes)
    return (center[0] - midpoint[0]) * normal[0] + (center[1] - midpoint[1]) * normal[1]


def _validate_edge_specs(
    specs: tuple[EdgeSpec, ...],
    edge_cells: dict[frozenset[str], list[tuple[str, tuple[str, str]]]],
) -> None:
    if any(len(spec.node_ids) != 2 for spec in specs):
        raise ValueError("edge specs must reference exactly two nodes")
    if len({spec.id for spec in specs}) != len(specs):
        raise ValueError("edge specs contain duplicate ids")
    if len({frozenset(spec.node_ids) for spec in specs}) != len(specs):
        raise ValueError("edge specs contain duplicate topological edges")
    if {frozenset(spec.node_ids) for spec in specs} != set(edge_cells):
        raise ValueError("edge specs do not cover the cell topology exactly")
    for spec in specs:
        key = frozenset(spec.node_ids)
        owners = edge_cells[key]
        if len(owners) > 2:
            raise ValueError(f"edge {spec.id} has more than two adjacent cells")
        owner_cells = {cell_id for cell_id, _ in owners}
        expected_cells = {cell for cell in (spec.left_cell, spec.right_cell) if cell is not None}
        if owner_cells != expected_cells:
            raise ValueError(f"edge {spec.id} adjacency does not match cell topology")
        if spec.left_cell is not None:
            left_orientations = [orientation for cell_id, orientation in owners if cell_id == spec.left_cell]
            if left_orientations != [spec.node_ids]:
                raise ValueError(f"edge {spec.id} direction must match the left cell orientation")


def _default_edge_specs(
    edge_cells: dict[frozenset[str], list[tuple[str, tuple[str, str]]]],
) -> tuple[EdgeSpec, ...]:
    specs: list[EdgeSpec] = []
    for index, (node_ids, owners) in enumerate(edge_cells.items()):
        if len(owners) > 2:
            raise ValueError(f"edge {sorted(node_ids)} has more than two adjacent cells")
        left_cell, oriented_node_ids = owners[0]
        right_cell = owners[1][0] if len(owners) == 2 else None
        specs.append(EdgeSpec(f"E{index}", oriented_node_ids, left_cell, right_cell))
    return tuple(specs)

File: libs/mesh/src/mesh.cpp
#include "mesh/mesh.hpp"

#include <algorithm>
#include <cmath>
#include <set>
#include <stdexcept>
#include <utility>

namespace scau::mesh {
namespace {

using Owner = std::pair<std::string, std::array<std::string, 2>>;
using EdgeOwners = std::unordered_map<std::string, std::vector<Owner>>;
using CellLookup = std::unordered_map<std::string, Cell>;

std::string edge_key(const std::array<std::string, 2>& node_ids) {
    if (node_ids[0] < node_ids[1]) {
        return node_ids[0] + "|" + node_ids[1];
    }
    return node_ids[1] + "|" + node_ids[0];
}

CellLookup cell_lookup(const std::vector<Cell>& cells) {
    CellLookup lookup;
    for (const auto& cell : cells) {
        lookup.emplace(cell.id, cell);
    }
    return lookup;
}

void validate_cells(const std::vector<Cell>& cells, const NodeLookup& nodes) {
    for (const auto& cell : cells) {
        if (cell.cell_type == CellType::Triangle && cell.node_count() != 3U) {
            throw std::invalid_argument("triangle " + cell.id + " must have 3 nodes");
        }
        if (cell.cell_type == CellType::Quadrilateral && cell.node_count() != 4U) {
            throw std::invalid_argument("quadrilateral " + cell.id + " must have 4 nodes");
        }
        std::set<std::string> unique_nodes(cell.node_ids.begin(), cell.node_ids.end());
        if (unique_nodes.size() != cell.node_count()) {
            throw std::invalid_argument("cell " + cell.id + " has duplicate nodes");
        }
        for (const auto& node_id : cell.node_ids) {
            if (!nodes.contains(node_id)) {
                throw std::invalid_argument("cell " + cell.id + " references unknown node " + node_id);
            }
        }
        if (cell_area(cell, nodes) == 0.0) {
            throw std::invalid_argument("cell " + cell.id + " has zero area");
        }
    }
}

EdgeOwners collect_edge_owners(const std::vector<Cell>& cells) {
    EdgeOwners edge_owners;
    for (const auto& cell : cells) {
        for (std::size_t index = 0; index < cell.edge_count(); ++index) {
            const std::array<std::string, 2> node_ids{
                cell.node_ids[index],
                cell.node_ids[(index + 1U) % cell.edge_count()],
            };
            edge_owners[edge_key(node_ids)].push_back({cell.id, node_ids});
        }
    }
    return edge_owners;
}

std::vector<EdgeSpec> default_edge_specs(const EdgeOwners& edge_owners) {
    std::vector<EdgeSpec> specs;
    specs.reserve(edge_owners.size());
    std::size_t index = 0;
    for (const auto& [key, owners] : edge_owners) {
        if (owners.size() > 2U) {
            throw std::invalid_argument("edge " + key + " has more than two adjacent cells");
        }
        specs.push_back(EdgeSpec{
            .id = "E" + std::to_string(index),
            .node_ids = owners[0].second,
            .left_cell = owners[0].first,
            .right_cell = owners.size() == 2U ? std::optional<std::string>{owners[1].first} : std::nullopt,
        });
        ++index;
    }
    return specs;
}

void validate_edge_specs(const std::vector<EdgeSpec>& specs, const EdgeOwners& edge_owners) {
    std::set<std::string> ids;
    std::set<std::string> topology;
    for (const auto& spec : specs) {
        if (!ids.insert(spec.id).second) {
            throw std::invalid_argument("edge specs contain duplicate ids");
        }
        if (!topology.insert(edge_key(spec.node_ids)).second) {
            throw std::invalid_argument("edge specs contain duplicate topological edges");
        }
    }
    std::set<std::string> expected_topology;
    for (const auto& [key, owners] : edge_owners) {
        expected_topology.insert(key);
    }
    if (topology != expected_topology) {
        throw std::invalid_argument("edge specs do not cover the cell topology exactly");
    }
    for (const auto& spec : specs) {
        const auto owners = edge_owners.at(edge_key(spec.node_ids));
        if (owners.size() > 2U) {
            throw std::invalid_argument("edge " + spec.id + " has more than two adjacent cells");
        }
        std::set<std::string> owner_cells;
        for (const auto& [cell_id, orientation] : owners) {
            owner_cells.insert(cell_id);
        }
        std::set<std::string> expected_cells;
        if (spec.left_cell.has_value()) {
            expected_cells.insert(*spec.left_cell);
        }
        if (spec.right_cell.has_value()) {
            expected_cells.insert(*spec.right_cell);
        }
        if (owner_cells != expected_cells) {
            throw std::invalid_argument("edge " + spec.id + " adjacency does not match cell topology");
        }
        if (spec.left_cell.has_value()) {
            const auto left_owner = std::find_if(owners.begin(), owners.end(), [&](const Owner& owner) {
                return owner.first == *spec.left_cell;
            });
            if (left_owner == owners.end() || left_owner->second != spec.node_ids) {
                throw std::invalid_argument("edge " + spec.id + " direction must match the left cell orientation");
            }
        }
    }
}

core::Real cell_side(
    const std::string& cell_id,
    const CellLookup& cells,
    const NodeLookup& nodes,
    const core::Vector2& midpoint,
    const core::Vector2& normal) {
    const auto center = cell_center(cells.at(cell_id), nodes);
    return ((center.x - midpoint.x) * normal.x) + ((center.y - midpoint.y) * normal.y);
}

core::Vector2 negate(core::Vector2 value) {
    return core::Vector2{.x = -value.x, .y = -value.y};
}

core::Vector2 normal_toward_cell(
    const std::optional<std::string>& cell_id,
    const CellLookup& cells,
    const NodeLookup& nodes,
    const core::Vector2& midpoint,
    core::Vector2 normal) {
    if (!cell_id.has_value()) {
        throw std::invalid_argument("boundary edge must reference one cell");
    }
    if (cell_side(*cell_id, cells, nodes, midpoint, normal) < 0.0) {
        return negate(normal);
    }
    return normal;
}

core::Vector2 normal_away_from_cell(
    const std::string& cell_id,
    const CellLookup& cells,
    const NodeLookup& nodes,
    const core::Vector2& midpoint,
    core::Vector2 normal) {
    if (cell_side(cell_id, cells, nodes, midpoint, normal) > 0.0) {
        return negate(normal);
    }
    return normal;
}

core::Vector2 normal_from_cell_to_cell(
    const std::string& left_cell,
    const std::string& right_cell,
    const CellLookup& cells,
    const NodeLookup& nodes,
    const core::Vector2& midpoint,
    core::Vector2 normal) {
    if (cell_side(right_cell, cells, nodes, midpoint, normal) < 0.0) {
        normal = negate(normal);
    }
    if (cell_side(left_cell, cells, nodes, midpoint, normal) > 0.0) {
        throw std::invalid_argument("edge normal does not separate left and right cells");
    }
    return normal;
}

core::Vector2 oriented_normal(
    const EdgeSpec& spec,
    const CellLookup& cells,
    const NodeLookup& nodes,
    core::Real length,
    core::Real dx,
    core::Real dy,
    const core::Vector2& midpoint) {
    const core::Vector2 normal{.x = dy / length, .y = -dx / length};
    if (!spec.left_cell.has_value()) {
        return normal_toward_cell(spec.right_cell, cells, nodes, midpoint, normal);
    }
    if (!spec.right_cell.has_value()) {
        return normal_away_from_cell(*spec.left_cell, cells, nodes, midpoint, normal);
    }
    return normal_from_cell_to_cell(*spec.left_cell, *spec.right_cell, cells, nodes, midpoint, normal);
}

std::vector<Edge> build_edges(
    const std::vector<Cell>& cells,
    const NodeLookup& nodes,
    const std::optional<std::vector<EdgeSpec>>& edge_specs) {
    const auto edge_owners = collect_edge_owners(cells);
    const auto specs = edge_specs.has_value() ? *edge_specs : default_edge_specs(edge_owners);
    const auto cells_by_id = cell_lookup(cells);
    validate_edge_specs(specs, edge_owners);

    std::vector<Edge> edges;
    edges.reserve(specs.size());
    for (const auto& spec : specs) {
        const auto& start = nodes.at(spec.node_ids[0]);
        const auto& end = nodes.at(spec.node_ids[1]);
        const auto dx = end.x - start.x;
        const auto dy = end.y - start.y;
        const auto length = std::hypot(dx, dy);
        if (length == 0.0) {
            throw std::invalid_argument("edge " + spec.id + " has zero length");
        }
        const core::Vector2 midpoint{.x = (start.x + end.x) * 0.5, .y = (start.y + end.y) * 0.5};
        edges.push_back(Edge{
            .id = spec.id,
            .node_ids = spec.node_ids,
            .left_cell = spec.left_cell,
            .right_cell = spec.right_cell,
            .normal = oriented_normal(spec, cells_by_id, nodes, length, dx, dy, midpoint),
            .length = length,
            .midpoint = midpoint,
        });
    }
    return edges;
}

std::vector<Node> control_nodes() {
    return {
        Node{.id = "N0", .x = 0.0, .y = 0.0},
        Node{.id = "N1", .x = 1.0, .y = 0.0},
        Node{.id = "N2", .x = 2.0, .y = 0.0},
        Node{.id = "N3", .x = 0.0, .y = 1.0},
        Node{.id = "N4", .x = 1.0, .y = 1.0},
        Node{.id = "N5", .x = 2.0, .y = 1.0},
    };
}

}  // namespace

std::size_t Cell::node_count() const noexcept {
    return node_ids.size();
}

std::size_t Cell::edge_count() const noexcept {
    return node_ids.size();
}

NodeLookup node_lookup(const std::vector<Node>& nodes) {
    NodeLookup lookup;
    for (const auto& node : nodes) {
        lookup.emplace(node.id, node);
    }
    return lookup;
}

Mesh build_mesh(std::vector<Node> nodes, std::vector<Cell> cells) {
    const auto nodes_by_id = node_lookup(nodes);
    validate_cells(cells, nodes_by_id);
    const auto edges = build_edges(cells, nodes_by_id, std::nullopt);
    return Mesh{.nodes = std::move(nodes), .cells = std::move(cells), .edges = edges};
}

Mesh build_mesh(std::vector<Node> nodes, std::vector<Cell> cells, std::vector<EdgeSpec> edge_specs) {
    const auto nodes_by_id = node_lookup(nodes);
    validate_cells(cells, nodes_by_id);
    const auto edges = build_edges(cells, nodes_by_id, std::move(edge_specs));
    return Mesh{.nodes = std::move(nodes), .cells = std::move(cells), .edges = edges};
}

Mesh build_mixed_minimal_mesh() {
    return build_mesh(
        control_nodes(),
        {
            Cell{.id = "C0", .cell_type = CellType::Quadrilateral, .node_ids = {"N0", "N1", "N4", "N3"}},
            Cell{.id = "C1", .cell_type = CellType::Triangle, .node_ids = {"N1", "N2", "N4"}},
            Cell{.id = "C2", .cell_type = CellType::Triangle, .node_ids = {"N2", "N5", "N4"}},
        },
        {
            EdgeSpec{.id = "E0", .node_ids = {"N0", "N1"}, .left_cell = std::nullopt, .right_cell = "C0"},
            EdgeSpec{.id = "E1", .node_ids = {"N1", "N4"}, .left_cell = "C0", .right_cell = "C1"},
            EdgeSpec{.id = "E2", .node_ids = {"N1", "N2"}, .left_cell = std::nullopt, .right_cell = "C1"},
            EdgeSpec{.id = "E3", .node_ids = {"N2", "N4"}, .left_cell = "C1", .right_cell = "C2"},
            EdgeSpec{.id = "E4", .node_ids = {"N4", "N3"}, .left_cell = "C0", .right_cell = std::nullopt},
            EdgeSpec{.id = "E5", .node_ids = {"N0", "N3"}, .left_cell = std::nullopt, .right_cell = "C0"},
            EdgeSpec{.id = "E6", .node_ids = {"N2", "N5"}, .left_cell = std::nullopt, .right_cell = "C2"},
            EdgeSpec{.id = "E7", .node_ids = {"N5", "N4"}, .left_cell = "C2", .right_cell = std::nullopt},
        });
}

Mesh build_quad_only_control_mesh() {
    return build_mesh(
        control_nodes(),
        {
            Cell{.id = "C0", .cell_type = CellType::Quadrilateral, .node_ids = {"N0", "N1", "N4", "N3"}},
            Cell{.id = "C1", .cell_type = CellType::Quadrilateral, .node_ids = {"N1", "N2", "N5", "N4"}},
        });
}

Mesh build_tri_only_control_mesh() {
    return build_mesh(
        control_nodes(),
        {
            Cell{.id = "C0", .cell_type = CellType::Triangle, .node_ids = {"N0", "N1", "N3"}},
            Cell{.id = "C1", .cell_type = CellType::Triangle, .node_ids = {"N1", "N4", "N3"}},
            Cell{.id = "C2", .cell_type = CellType::Triangle, .node_ids = {"N1", "N2", "N4"}},
            Cell{.id = "C3", .cell_type = CellType::Triangle, .node_ids = {"N2", "N5", "N4"}},
        });
}

core::Real cell_area(const Cell& cell, const NodeLookup& nodes) {
    core::Real twice_area = 0.0;
    for (std::size_t index = 0; index < cell.node_count(); ++index) {
        const auto& current = nodes.at(cell.node_ids[index]);
        const auto& following = nodes.at(cell.node_ids[(index + 1U) % cell.node_count()]);
        twice_area += (current.x * following.y) - (following.x * current.y);
    }
    return std::abs(twice_area) * 0.5;
}

core::Vector2 cell_center(const Cell& cell, const NodeLookup& nodes) {
    core::Real signed_area_twice = 0.0;
    core::Real centroid_x = 0.0;
    core::Real centroid_y = 0.0;
    for (std::size_t index = 0; index < cell.node_count(); ++index) {
        const auto& current = nodes.at(cell.node_ids[index]);
        const auto& following = nodes.at(cell.node_ids[(index + 1U) % cell.node_count()]);
        const auto cross = (current.x * following.y) - (following.x * current.y);
        signed_area_twice += cross;
        centroid_x += (current.x + following.x) * cross;
        centroid_y += (current.y + following.y) * cross;
    }
    if (signed_area_twice == 0.0) {
        throw std::invalid_argument("cell " + cell.id + " has zero area");
    }
    return core::Vector2{.x = centroid_x / (3.0 * signed_area_twice), .y = centroid_y / (3.0 * signed_area_twice)};
}

}  // namespace scau::mesh
